architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,446 @@
+    - Check if the template contains a `param application`. If it does, the controller will query the Radius API to get an application ID (must have exactly one application if left unspecified like this) and deploy the template using that application.
+2. The controller will send the ARM JSON to the Radius API to deploy the resources. If there are any errors during the deployment, the controller will update the `status.phrase` and the `status.description` of the BicepDeployment resource to indicate that the deployment has failed and include the error message.
+
+> TODO: We need to figure out deletion of specific resources within the template. Let's discuss in the design meeting.
+
+### Architecture Diagram
+<!--
+Provide a diagram of the system architecture, illustrating how different
+components interact with each other in the context of this proposal.
+
+Include separate high level architecture diagram and component specific diagrams, wherever appropriate.
+-->
+
+![Architecture Diagram](2024-08-bicepdeployment-controller/architecture.png)
+
+### Detailed Design
+
+<!--
+This section should be detailed and thorough enough that another developer
+could implement your design and provide enough detail to get a high confidence
+estimate of the cost to implement the feature but isn’t as detailed as the 
+code. Be sure to also consider testability in your design.
+
+For each change, give each "change" in the proposal its own section and
+describe it in enough detail that someone else could implement it. Cover
+ALL of the important decisions like names. Your goal is to get an agreement
+to proceed with coding and PRs.
+
+If there are alternatives you are considering please include that in the open
+questions section. If the product has a layered architecture, it's good to
+align these sections with the product's layers. This will help readers use
+their current understanding to understand your ideas.
+
+Discuss the rationale behind architectural choices and alternative options 
+considered during the design process.
+-->
+
+#### BicepDeployment Custom Resource Definition
+
+The `BicepDeployment` resource will be a new Kubernetes CRD that will be used to contain the Bicep template and parameters for deploying Radius resources. The CRD will have the following fields:
+
+```go
+// BicepDeploymentSpec defines the desired state of an BicepDeployment
+type BicepDeploymentSpec struct {
+  // Template is the Bicep template that defines the resources to deploy.
+  // +kubebuilder:validation:Required
+  Template string `json:"template"`
+
+  // Parameters is the Bicep parameters file that specifies the parameters for the template.
+  // +kubebuilder:validation:Optional
+  Parameters string `json:"parameters"`
```

Comment:
Do we need to support more than one of these?

Reply:
updated the design to support multiple param files

Reply:
So the inputs can be multiple files, but they are stored on spec as a single file?

Reply:
It's clear from the example now that this stores a stringified parameter file (JSON).